Camel metrics JSON serialization throws ClassNotFoundException

Camel metrics not working in release 5.0. 

```
2017-12-14 14:16:23,354 INFO  [org.apache.camel.spring.SpringCamelContext] (MSC service thread 1-2) Apache Camel 2.20.1 (CamelContext: spring-context) is shutdown in 0.007 seconds
2017-12-14 14:16:23,355 ERROR [org.jboss.msc.service.fail] (MSC service thread 1-2) MSC000001: Failed to start service jboss.deployment.unit."avs-ws-client.war".INSTALL: org.jboss.msc.service.StartException in service jboss.deployment.unit."avs-ws-client.war".INSTALL: WFLYSRV0153: Failed to process phase INSTALL of deployment "avs-ws-client.war"
	at org.jboss.as.server.deployment.DeploymentUnitPhaseService.start(DeploymentUnitPhaseService.java:172)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:2032)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1955)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NoClassDefFoundError: com/fasterxml/jackson/core/JsonProcessingException
	at org.apache.camel.component.metrics.messagehistory.MetricsMessageHistoryFactory.doStart(MetricsMessageHistoryFactory.java:155)
	at org.apache.camel.support.ServiceSupport.start(ServiceSupport.java:61)
	at org.apache.camel.util.ServiceHelper.startService(ServiceHelper.java:75)
	at org.apache.camel.impl.DefaultCamelContext.doAddService(DefaultCamelContext.java:1495)
	at org.apache.camel.impl.DefaultCamelContext.addService(DefaultCamelContext.java:1455)
	at org.apache.camel.impl.DefaultCamelContext.doStartCamel(DefaultCamelContext.java:3394)
	at org.apache.camel.impl.DefaultCamelContext.access$000(DefaultCamelContext.java:208)
	at org.apache.camel.impl.DefaultCamelContext$2.call(DefaultCamelContext.java:3249)
	at org.apache.camel.impl.DefaultCamelContext$2.call(DefaultCamelContext.java:3245)
	at org.apache.camel.impl.DefaultCamelContext.doWithDefinedClassLoader(DefaultCamelContext.java:3268)
	at org.apache.camel.impl.DefaultCamelContext.doStart(DefaultCamelContext.java:3245)
	at org.apache.camel.support.ServiceSupport.start(ServiceSupport.java:61)
	at org.apache.camel.impl.DefaultCamelContext.start(DefaultCamelContext.java:3168)
	at org.apache.camel.spring.SpringCamelContext.start(SpringCamelContext.java:133)
	at org.wildfly.extension.camel.deployment.CamelContextActivationProcessor$1.call(CamelContextActivationProcessor.java:61)
	at org.wildfly.extension.camel.deployment.CamelContextActivationProcessor$1.call(CamelContextActivationProcessor.java:58)
	at org.wildfly.extension.camel.proxy.ProxyUtils$1.invoke(ProxyUtils.java:51)
	at com.sun.proxy.$Proxy196.call(Unknown Source)
	at org.wildfly.extension.camel.proxy.ProxyUtils.invokeProxied(ProxyUtils.java:55)
	at org.wildfly.extension.camel.deployment.CamelContextActivationProcessor.deploy(CamelContextActivationProcessor.java:58)
	at org.jboss.as.server.deployment.DeploymentUnitPhaseService.start(DeploymentUnitPhaseService.java:165)
	... 5 more
Caused by: java.lang.ClassNotFoundException: com.fasterxml.jackson.core.JsonProcessingException from [Module "org.apache.camel.component.metrics" from local module loader @17d99928 (finder: local module finder @3834d63f (roots: C:\jboss\wildfly-11.0.0.Final\modules,C:\jboss\wildfly-11.0.0.Final\modules\system\layers\fuse,C:\jboss\wildfly-11.0.0.Final\modules\system\layers\base))]
	at org.jboss.modules.ModuleClassLoader.findClass(ModuleClassLoader.java:198)
	at org.jboss.modules.ConcurrentClassLoader.performLoadClassUnchecked(ConcurrentClassLoader.java:412)
	at org.jboss.modules.ConcurrentClassLoader.performLoadClass(ConcurrentClassLoader.java:400)
	at org.jboss.modules.ConcurrentClassLoader.loadClass(ConcurrentClassLoader.java:116)
	... 26 more

```

In order to fix the issue the following files had to be modified.

Added:

	<module name="com.fasterxml.jackson.core.jackson-core" />

To:

modules\system\layers\fuse\io\dropwizard\metrics\main\module.xml
modules\system\layers\fuse\org\apache\camel\component\metrics\main\module.xml

--- a/feature/modules/etc/smartics/camel-modules.xml
+++ b/feature/modules/etc/smartics/camel-modules.xml
@@ -1833,6 +1833,7 @@
             <include module="org.springframework.boot" />
         </apply-to-dependencies>
         <dependencies>
+            <module name="com.fasterxml.jackson.core.jackson-core"/>
             <module name="io.dropwizard.metrics" export="true"/>
             <module name="org.slf4j"/>
         </dependencies>

--- a/feature/modules/etc/smartics/other-modules.xml
+++ b/feature/modules/etc/smartics/other-modules.xml
@@ -738,6 +738,7 @@
         <include artifact="io.dropwizard.metrics:metrics-core" />
         <include artifact="io.dropwizard.metrics:metrics-json" />
         <dependencies>
+            <module name="com.fasterxml.jackson.core.jackson-core" />
             <module name="javax.api" />
         </dependencies>
     </module>

--- a/feature/modules/src/main/resources/modules/system/layers/fuse/io/dropwizard/metrics/main/module.xml
+++ b/feature/modules/src/main/resources/modules/system/layers/fuse/io/dropwizard/metrics/main/module.xml
@@ -5,6 +5,7 @@
     <artifact name="${io.dropwizard.metrics:metrics-json}" />
   </resources>
   <dependencies>
+    <module name="com.fasterxml.jackson.core.jackson-core" />
     <module name="javax.api" />
     <module name="com.fasterxml.jackson.core.jackson-databind" />
     <module name="org.slf4j" />

--- a/feature/modules/src/main/resources/modules/system/layers/fuse/org/apache/camel/component/metrics/main/module.xml
+++ b/feature/modules/src/main/resources/modules/system/layers/fuse/org/apache/camel/component/metrics/main/module.xml
@@ -4,6 +4,7 @@
     <artifact name="${org.apache.camel:camel-metrics}" />
   </resources>
   <dependencies>
+    <module name="com.fasterxml.jackson.core.jackson-core" />
     <module name="io.dropwizard.metrics" export="true" />
     <module name="org.slf4j" />
     <module name="com.fasterxml.jackson.core.jackson-databind" />

--- a/itests/standalone/basic/src/test/java/org/wildfly/camel/test/metrics/MetricsIntegrationTest.java
+++ b/itests/standalone/basic/src/test/java/org/wildfly/camel/test/metrics/MetricsIntegrationTest.java
@@ -21,9 +21,14 @@ package org.wildfly.camel.test.metrics;
 
 import java.util.SortedMap;
 
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.MetricRegistry;
+
 import org.apache.camel.CamelContext;
 import org.apache.camel.ProducerTemplate;
 import org.apache.camel.builder.RouteBuilder;
+import org.apache.camel.component.metrics.messagehistory.MetricsMessageHistoryFactory;
+import org.apache.camel.component.metrics.messagehistory.MetricsMessageHistoryService;
 import org.apache.camel.impl.DefaultCamelContext;
 import org.apache.camel.impl.SimpleRegistry;
 import org.jboss.arquillian.container.test.api.Deployment;
@@ -34,9 +39,6 @@ import org.junit.Assert;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.wildfly.extension.camel.CamelAware;
-
-import com.codahale.metrics.Counter;
-import com.codahale.metrics.MetricRegistry;
 
 @CamelAware
 @RunWith(Arquillian.class)
@@ -54,15 +56,15 @@ public class MetricsIntegrationTest {
         registry.put("metricRegistry", metricRegistry);
 
         CamelContext camelctx = new DefaultCamelContext(registry);
-        try {
-            camelctx.addRoutes(new RouteBuilder() {
-                @Override
-                public void configure() throws Exception {
-                    from("direct:start")
-                    .to("metrics:counter:simple.counter?increment=5");
-                }
-            });
+        camelctx.addRoutes(new RouteBuilder() {
+            @Override
+            public void configure() throws Exception {
+                from("direct:start")
+                .to("metrics:counter:simple.counter?increment=5");
+            }
+        });
 
+        try {
             camelctx.start();
 
             ProducerTemplate template = camelctx.createProducerTemplate();
@@ -70,7 +72,7 @@ public class MetricsIntegrationTest {
 
             SortedMap<String, Counter> counters = metricRegistry.getCounters();
             Counter counter = null;
-            for(String counterName : counters.keySet()) {
+            for (String counterName : counters.keySet()) {
                 if (counterName.equals("simple.counter")) {
                     counter = counters.get(counterName);
                     break;
@@ -79,6 +81,35 @@ public class MetricsIntegrationTest {
 
             Assert.assertNotNull("Counter simple.counter was null", counter);
             Assert.assertEquals(5, counter.getCount());
+        } finally {
+            camelctx.stop();
+        }
+    }
+
+    @Test
+    public void testMetricsJsonSerialization() throws Exception {
+        SimpleRegistry registry = new SimpleRegistry();
+        MetricRegistry metricRegistry = new MetricRegistry();
+        registry.put("metricRegistry", metricRegistry);
+
+        MetricsMessageHistoryFactory messageHistoryFactory = new MetricsMessageHistoryFactory();
+        messageHistoryFactory.setMetricsRegistry(metricRegistry);
+
+        CamelContext camelctx = new DefaultCamelContext(registry);
+        camelctx.setMessageHistoryFactory(messageHistoryFactory);
+        camelctx.addRoutes(new RouteBuilder() {
+            @Override
+            public void configure() throws Exception {
+                from("direct:start")
+                .to("metrics:counter:simple.counter?increment=5");
+            }
+        });
+
+        camelctx.start();
+        try {
+            MetricsMessageHistoryService service = camelctx.hasService(MetricsMessageHistoryService.class);
+            String json = service.dumpStatisticsAsJson();
+            Assert.assertTrue(json.contains("\"gauges\":{},\"counters\":{},\"histograms\":{},\"meters\":{},\"timers\":{}"));
         } finally {
             camelctx.stop();
         }

